src/cargo/core/compiler/job_queue.rs

```diff
@@ -306,12 +381,31 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
                 self.run(&unit, job, cx, scope)?;
             }
 
+            info!(
+                "tokens: {}, rustc_tokens: {}, waiting_rustcs: {}",
+                tokens.len(),
+                rustc_tokens.len(),
+                to_send_clients.len()
+            );
+
             // If after all that we're not actually running anything then we're
             // done!
             if self.active.is_empty() {
                 break;
             }
 
+            // If we managed to acquire some extra tokens, send them off to a waiting rustc.
+            let extra_tokens = tokens.len() - (self.active.len() - 1);
```

Comment:
If you don't mind doing so, I think that we need to refactor this loop within Cargo to be a bit more readable (it's starting to get a bit sprawling).

I'm imagining something like:

```rust
let mut state = RunState::new(self);
loop {
    state.spawn_work_if_possible()?;
    if state.finished() { break; }
    state.grant_rustc_token_requests()?;
    for event in state.wait_for_events()? {
        state.process_event(event)?;
    }
}
```

Basically I'd like to make the structure of the loop much clearer and then have clear documentation for each method what happens where. This would also hopefully lend itself to better-documented fields as well as easily adding debug assertions if necessary.